### PR TITLE
Separate php.ini settings from fpm settings

### PR DESCRIPTION
**Pull request by minorOffense**
Alright, so this is going to need some testing. I added a new variable to store the location of php.ini for each OS. Everything except Redhat was an "educated guess" as to the proper location.

The new "php::global" class runs in the init.pp file and sets values in the file (similarly to php-cli.ini)

The fpm settings have been moved into a php-fpm.ini file which according to the PHP docs should be autoloaded based on the package name (exactly the same way php-cli.ini works).

I provisioned a machine and tested the config. Appears to work and nothing breaks. But like I said, I haven't been able to test the other OSes yet.

### CODE REVIEW

**Comment by minorOffense** [comment]
Working on https://github.com/mayflower/puppet-php/issues/78

**Comment by minorOffense** [comment]
Well there's work to do. Any hints you might be able to toss my way to figure out why the tests broke would be great.

**Comment by fpletz** [comment]
Looking good after a quick glance if the behavior is indeed as stated in your link in #78. Will test this in a few days and get back to you.

We don't currently have a good test coverage. As the `php::config` resource already has specs, I think we can skip those for your new class.

**Comment by minorOffense** [comment]
Doing some of my own testing, I noticed a warning on one of our vms

``
==> default: Could not retrieve fact='phpversion', resolution='<anonymous>': undefined method `split' for nil:NilClass
``

Not sure if it's related or not. We've been messing with installing newer versions of PHP on centos. But if you run into this too then odds are it's due to the patch.

**Comment by minorOffense** [comment]
Just checking in to see if you need anything from me. Want to help with the patch as much as possible.

**Comment by codec** [comment]
@minorOffense This is thanks to my :hankey: code in `lib/facter/phpversion.rb`. It's probably only happening on the first run?

**Comment by minorOffense** [comment]
Yeah it's only on the first run. Ok so that's not related. Good to know ;-)

**Comment by globin** [comment]
Docs say that it will always fail on first run.
https://github.com/mayflower/puppet-php#facts

File: manifests/init.pp [hunk 1]
```diff
@@ -76,6 +76,13 @@
     } ->
   anchor { 'php::end': }
 
+  # Configure global PHP settings in php.ini
+  Anchor['php::begin'] ->
+  class {'::php::global':
+    settings => $real_settings,
+  } ->
+  Anchor ['php::end']
```
Comment (codec) [inline manifests/init.pp]: @minorOffense The whitespace is causing the specs to fail using future parser. :see_no_evil:

**Comment by mmoll** [comment]
@minorOffense could you rebase onto current master?

**Comment by minorOffense** [comment]
Merged in changes from master but didn't fix that whitespace thing yet.

**Comment by fpletz** [comment]
Squashed, rebased and pushed to master with the whitespace fix. Thanks!

**Comment by voxel01** [comment]
The fix breaks backward compatibitilty. As soon as you use php::settings in hiera you get an compileerror:
```Error: Could not retrieve catalog from remote server: Error 400 on SERVER: Duplicate declaration: Php::Config::Setting[/etc/php.ini: Date/date.timezone] is already declared; cannot redeclare at /etc/puppet/modules/common/php/manifests/config.pp:32 on node client.puppet.local```

The error was raised when I added the follwing config to hiera file. Which was correct until now.
```
php::settings:
    "Date/date.timezone": "Europe/Vienna"
```

**Comment by mmoll** [comment]
4.0.0 is a major release because of all the backward-incompatibilities.

**Comment by mmoll** [comment]
if the docs need updating, please open a PR :)

**Pull request closed without merging**